lib/babe/verify.go

```diff
@@ -344,29 +344,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 	// hashes is hashes of all blocks with same block number as header.Number
 	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
 
-	for _, hash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(hash)
+	for _, currentHash := range hashes {
+		currentHeader, err := b.blockState.GetHeader(currentHash)
 		if err != nil {
+			logger.Errorf("failed get header %s", err)
 			continue
 		}
 
 		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
 		if err != nil {
+			logger.Errorf("failed to get authority index %s", err)
 			continue
 		}
 
+		if len(currentHeader.Digest.Types) == 0 {
+			logger.Errorf("current header missing digest")
+		}
+
+		currentPreDigestItem := currentHeader.Digest.Types[0]
+		currentPreDigest, ok := currentPreDigestItem.Value().(types.PreRuntimeDigest)
+		if !ok {
+			return fmt.Errorf("%w: got %T", types.ErrNoFirstPreDigest, currentPreDigestItem.Value())
+
+		}
+
+		currentBabePreDigest, err := b.verifyPreRuntimeDigest(&currentPreDigest)
+		if err != nil {
+			return fmt.Errorf("failed to verify pre-runtime digest: %w", err)
+		}
+
+		var isCurrentBlockProducerPrimary bool
+		switch currentBabePreDigest.(type) {
+		case types.BabePrimaryPreDigest:
+			isCurrentBlockProducerPrimary = true
+		default:
+		}
+
+		var isExistingBlockProducerPrimary bool
 		var existingBlockProducerIndex uint32
 		switch d := babePreDigest.(type) {
 		case types.BabePrimaryPreDigest:
 			existingBlockProducerIndex = d.AuthorityIndex
+			isExistingBlockProducerPrimary = true
 		case types.BabeSecondaryVRFPreDigest:
 			existingBlockProducerIndex = d.AuthorityIndex
 		case types.BabeSecondaryPlainPreDigest:
 			existingBlockProducerIndex = d.AuthorityIndex
 		}
 
-		// same authority won't produce two different blocks at the same block number
-		if currentBlockProducerIndex == existingBlockProducerIndex && hash != header.Hash() {
+		// same authority won't produce two different blocks at the same block number as primary block producer
+		if currentBlockProducerIndex == existingBlockProducerIndex &&
+			currentHash.Equal(header.Hash()) &&
```

Comment:
should be not equal, right?
```suggestion
			!currentHash.Equal(header.Hash()) &&
```